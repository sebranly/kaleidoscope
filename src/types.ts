export enum Color {
  Yellow = 'yellow',
  Green = 'green',
  Blue = 'blue',
  Violet = 'violet',
  Orange = 'orange',
  Red = 'red',
  Pink = 'pink',
  White = 'white'
}

export type Episode = {
  color: Color;
  defaultNumber: number;
  director?: string;
  hoursFromHeist: number;
  title: string;
  writers: string[];
};

export type TimeUnits = {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
};

export enum TimeLabel {
  Day = 'day',
  Hour = 'hour',
  Minute = 'minute',
  Second = 'second'
}

export enum Direction {
  Up = 'up',
  Down = 'down'
}
