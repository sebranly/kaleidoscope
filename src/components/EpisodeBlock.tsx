import * as React from 'react';
import { EPISODE_COUNT } from '../constants/general';
import { Direction, Episode } from '../types';
import { pluralize, swapEpisodes } from '../utils';

export interface EpisodeBlockProps {
  episode: Episode;
  index: number;
  episodesList: Episode[];
  setEpisodesList: React.Dispatch<React.SetStateAction<Episode[]>>;
}

const EpisodeBlock: React.FC<EpisodeBlockProps> = (props) => {
  const { episode, episodesList, index, setEpisodesList } = props;
  const { color, defaultNumber, director, title, writers } = episode;

  const writersCopy = pluralize('Writer', writers.length);
  const isFirst = index === 0;
  const isLast = index === EPISODE_COUNT - 1;

  const onClick = (direction: Direction) => {
    const episodesListBis = swapEpisodes(episodesList, direction, index);

    setEpisodesList(episodesListBis);
  };

  return (
    <div className="episode-block">
      <div className="episode-header">
        {!isFirst && (
          <button className="episode-arrow episode-up" onClick={() => onClick(Direction.Up)}>
            ðŸ”¼
          </button>
        )}
        <>{title}</>
        {!isLast && (
          <button className="episode-arrow episode-down" onClick={() => onClick(Direction.Down)}>
            ðŸ”½
          </button>
        )}
      </div>
      <div className={`episode-content-separator bg-color-${color}`}></div>
      <div className="episode-content">
        <div className="episode-director">
          <b>Director:</b> {director}
        </div>
        <div className="episode-writers">
          <b>{writersCopy}:</b> {writers.join(', ')}
        </div>
        <div className="episode-number">
          <b>Episode</b> {defaultNumber}
        </div>
      </div>
    </div>
  );
};

export { EpisodeBlock };
