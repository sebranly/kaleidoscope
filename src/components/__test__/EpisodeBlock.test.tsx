import * as React from 'react';
import { render } from '@testing-library/react';

import { EpisodeBlock } from '../EpisodeBlock';
import { Color, Episode } from '../../types';

test('EpisodeBlock', () => {
  const episode: Episode = {
    color: Color.Red,
    defaultNumber: 1,
    director: 'TBA',
    hoursFromHeist: -24,
    title: 'Some title',
    writers: ['Someone', 'Someone else']
  };

  const fn = jest.fn();
  const createComponent = () => (
    <EpisodeBlock episode={episode} index={1} episodesList={[episode]} setEpisodesList={fn} />
  );

  const { container } = render(createComponent());
  expect(container.childNodes).toMatchSnapshot();
});
