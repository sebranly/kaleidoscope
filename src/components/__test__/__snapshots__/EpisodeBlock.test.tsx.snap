// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeBlock 1`] = `
NodeList [
  <div
    class="episode-block"
  >
    <div
      class="episode-ribbon-number1-index1 episode-ribbon bg-color-red"
    />
    <div
      class="episode-ribbon-number2-index1 episode-ribbon bg-color-red"
    />
    <div
      class="episode-ribbon-number3-index1 episode-ribbon bg-color-red"
    />
    <div
      class="episode-content"
    >
      <div
        class="episode-header"
      >
        <button
          class="episode-arrow episode-up"
        >
          ðŸ”¼
        </button>
        Some title
        <button
          class="episode-arrow episode-down"
        >
          ðŸ”½
        </button>
      </div>
      <div
        class="episode-content-separator bg-color-red"
      />
      <div
        class="episode-details"
      >
        <div
          class="episode-writers"
        >
          <b>
            Writers
            :
          </b>
           
          Someone, Someone else
        </div>
        <div
          class="episode-director"
        >
          <b>
            Director:
          </b>
           
          TBA
        </div>
        <div
          class="episode-number"
        >
          <b>
            Episode:
          </b>
           #
          1
        </div>
      </div>
    </div>
  </div>,
]
`;
